### PR TITLE
[SPARK-24782][SQL] Simplify conf retrieval in SQL expressions

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

The PR simplifies the retrieval of config in `size`, as we can access them from tasks too thanks to SPARK-24250.

## How was this patch tested?

existing UTs

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

**Comment by mgaido91** [comment]
cc @cloud-fan @MaxGekk

**Comment by mgaido91** [comment]
retest this please

**Comment by cloud-fan** [comment]
@MaxGekk The `DataFrame` contains an analyzed query, so the `size` expression in that df is fixed and will not change with the config. I think this is an existing behavior for many analyze phase configs.

**Comment by cloud-fan** [comment]
@mgaido91 can you check if there are more places we can simplify? AFAIK there are many places like this.

**Comment by mgaido91** [comment]
@cloud-fan sure, but do you think it is a good idea to do that in a followup? Wouldn't be creating a minor ticket or a separate PR for that a bit cleaner? Thanks.

**Comment by MaxGekk** [comment]
@cloud-fan Just for testing, I changed implementation slightly:
- removed `@transient lazy val legacySizeOfNull = SQLConf.get.legacySizeOfNull`
- and call the `SQLConf.get.legacySizeOfNull` directly in 3 places instead of checking the lazy val.

```
scala> val df = spark.sql("select size(null)")
scala> spark.conf.set("spark.sql.legacy.sizeOfNull", "false")
scala> df.show()
+----------+
|size(NULL)|
+----------+
|      null|
+----------+
```

**Comment by mgaido91** [comment]
@MaxGekk you can check that even without my change, ie. with the code of current master coming from your PR, the result is the same (you can check on your own). The reason is what was said by @cloud-fan which I try to elaborate a bit more here.

During analysis/optimization, deterministic operations/functions are executed and they are replaced in the plan with their result. As the analysis/optimization is performed only once when the dataframe is created, if you reuse the same dataframe multiple times you will always get the same result, as it is stored in the plan. If you want new configs to take effect you need to create a new dataframe.

**Comment by cloud-fan** [comment]
@mgaido91 we can create a JIRA for it and this PR doesn't need to be a minor followup.

**Comment by MaxGekk** [comment]
@mgaido91 I am not blaming you broke the current implementation. I am just testing yours. For example, now it is not clear for me why do you need `lazy val` instead of just `val` in `@transient lazy val legacySizeOfNull = SQLConf.get.legacySizeOfNull` and `@transient` if `legacySizeOfNull` must be instantiated only once at analyzed phase?

**Comment by HyukjinKwon** [comment]
It's nice to test someone's PR and changes but better leave only some related comments valid for the change to make people less confused.

**Comment by mgaido91** [comment]
@MaxGekk the main reason is to avoid it to be serialized to the executors, so we use the `@transient`. The main reason for using lazy is to get the value only if needed. Probably it is not a big deal to get rid of lazy.

@cloud-fan thanks, then I'll create a JIRA. Thanks.

**Comment by MaxGekk** [comment]
> Probably it is not a big deal to get rid of lazy.

Sure. You just do unnecessary synchronization inside of lazy implementation when you read the `lazy val` per each `null` input, for example here: https://github.com/apache/spark/pull/21736/files#diff-9853dcf5ce3d2ac1e94d473197ff5768R103 . 

In general, the difference between current implementation and yours, you just moved constructor's parameter to its body at the beginning and added two key words `lazy` and `@transient`. Basically this PR propose only those 3 changes. Moving constructor parameter to its body is understandable - you simplified tests. For `lazy`, you just degrade performance in some cases. `@transient` - serialization of `Boolean` value - not big deal. 

As far as I see you are going to repeat the same pattern in another places (functions). Maybe it makes sense to make the pattern clear here before repeating it?

**Comment by mgaido91** [comment]
Yes, I agree on choosing the right pattern. I think `@transient` is not a big harm to be kept. As far as `lazy` is regarded, I think that the performance problem introduced is not very important, as we are using it only we times (it is indeed used several times in the interpreted execution, but we don't use interpreted execution unless in very rare cases). Anyway, I am fine removing `lazy` and keeping `@transient`. What do you think?

**Comment by cloud-fan** [comment]
@MaxGekk If we call `SQLConf.get.legacySizeOfNull` directly in the execution path, we can update the result for an existing df, but not always. If the physical plan of the df is also materiazed, then the generated code is fixed and the updated config values can't be respected.

Regarding the `@transient lazy val`, if it doesn't have much benefit, shall we just remove it for simplicity?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -89,15 +89,9 @@ trait BinaryArrayExpressionWithImplicitCast extends BinaryExpression
...
-    child: Expression,
-    legacySizeOfNull: Boolean)
-  extends UnaryExpression with ExpectsInputTypes {
+case class Size(child: Expression) extends UnaryExpression with ExpectsInputTypes {
 
-  def this(child: Expression) =
-    this(
-      child,
-      legacySizeOfNull = SQLConf.get.getConf(SQLConf.LEGACY_SIZE_OF_NULL))
+  val legacySizeOfNull = SQLConf.get.legacySizeOfNull
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Actually we can do this even without my executor-side SQLConf change: the conf value will be determined at driver side during expression creation and serialized/sent to executors...

**Comment by cloud-fan** [comment]
LGTM

**Review by gatorsmile — COMMENTED** [review]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**